sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java

```diff
@@ -430,12 +430,15 @@ public int numValidRows() {
   /**
    * Sets (replaces) the column at `ordinal` with column. This can be used to do very efficient
    * projections.
+   * Note: Caller must ensure that ColumnVector in column[ordinal] is not used in other places
    */
   public void setColumn(int ordinal, ColumnVector column) {
-    if (column instanceof OffHeapColumnVector) {
-      throw new UnsupportedOperationException("Need to ref count columns.");
-    }
+    ColumnVector oldColumn = columns[ordinal];
     columns[ordinal] = column;
+    if (oldColumn instanceof OffHeapColumnVector) {
```

Comment:
we can call `oldColumn.close()` directly, as `OnHeapColumnVector.close` is no-op